File: Qytetet_Ruby/lib/especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "jugador.rb"
module ModeloQytetet
  class Especulador < Jugador

    def initialize(jugador, fianza)
      super(jugador.nombre);
      @encarcelado = jugador.encarcelado
      @saldo = jugador.saldo
      @carta_libertad = jugador.carta_libertad
      @casilla_actual = jugador.casilla_actual
      @propiedades = jugador.propiedades
      @factor_especulacion = 2
      @fianza = fianza
    end
    
    def convertirme(fianza) 
        return self
    end
    
    def pagar_impuestos(cantidad)
      modificar_saldo(-cantidad / 2)
    end
    
    def pagar_fianza(cantidad)
      pagada = false
      
      if(saldo >= cantidad)
        modificar_saldo(-cantidad)
        pagada = true
      end
      
      pagada
    end
    def ir_a_carcel(casilla)
      if(!pagar_fianza(@fianza))
        @casilla_actual = casilla
        @encarcelado = true
      end
    end 
    protected :pagar_impuestos
    private :pagar_fianza
  end
end

File: QytetetJava/src/modeloqytetet/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author antonio
 */
public class Especulador extends Jugador {

    private int fianza;
    static int FactorEspeculador = 2;
    /**
     *
     * @param jugador
     * @param fianza
     */
    protected Especulador(Jugador jugador, int fianza) {
        super(jugador);
        this.fianza = fianza;
    }

    /**
     *
     * @param cantidad
     */
    @Override
    protected void pagarImpuestos(int cantidad) {
        super.pagarImpuestos(cantidad / 2);
    }

    /**
     *
     * @param casilla
     */
    @Override
    protected void irACarcel(Casilla casilla) {
        if (!pagarFianza(fianza)) {
            super.irACarcel(casilla);
        }
    }

    /**
     *
     * @param fianza
     * @return
     */
    @Override
    protected Especulador convertirme(int fianza) {
        return this;
    }

    private boolean pagarFianza(int cantidad) {
        boolean pagada = false;
        if (super.getSaldo() >= cantidad) {
            super.modificarSaldo(-cantidad);
            pagada = true;
        }
        return pagada;
    }

    public static int getFactorEspeculador() {
        return Especulador.FactorEspeculador;
    }

    public int getFianza() {
        return fianza;
    }
    
    

}
